server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -132,6 +133,18 @@ private void handlePrimaryResult(final PrimaryResultT primaryResult) {
             if (logger.isTraceEnabled()) {
                 logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
             }
+            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
+
+            pendingActions.incrementAndGet();
+            replicasProxy.onPrimaryOperationComplete(
+                replicaRequest,
+                replicationGroup.getRoutingTable(),
+                ActionListener.wrap(
+                    ignored -> decPendingAndFinishIfNeeded(),
+                    exception -> this.finishAsFailed(new OnPrimaryOperationCompleteException(primary.routingEntry().shardId(), exception))
```

Comment:
Hmm I don't think we should `finishAsFailed` here - we should treat failures here similarly to replica failures, which involves adding them to `shardReplicaFailures` but allowing the overall operation to succeed.

Reply:
Hi @DaveCTurner , I do not think we can do that, as there is no way we can know which shard/node failed. `onPrimaryOperationComplete()` is a general customization function that does not operate on a single shard/replica, but the user is able to do whatever they would like. Thus, I believe the `finishAsFailed` is better. 

> but allowing the overall operation to succeed.

I thought the overall strategy we wanted was to fail the refresh if the unpromotable refresh fails. So this is another reason I chose `finishAsFailed`.

If we would like it to still make progress, then we could:
* Either revert my latest commit, which would bubble up the previous exception and the primary would retry with success (after removing the disconnected node), but the Refresh would then succeed.
* We could make more infrastructure to somehow make `onPrimaryOperationComplete()` on a per-shard level. But then we go back to my idea of integrating the unpromotable replication inside ReplicationOperation rather than leaving it to the implementor to do that outside of it.

Please advise :)

Reply:
I believe we can create a `org.elasticsearch.action.support.replication.ReplicationResponse.ShardInfo.Failure` representing the failure - the `nodeId` field is (or at least claims to be) optional.

If not, I'd rather just log and drop the exception for now and come back to the question of failure-handling in a follow-up.

Reply:
You're right, thanks for the tip! I managed to make it behave like a replica shard failure. Feel free to see the relevant commit when it's pushed and comment :)